File: truth_set/go/structure/organisation/duplication/UserService.go
package userservice

import (
	"fmt"
	"math/rand"
	"time"
)

type User struct {
	ID             string
	Name           string
	Age            int
	IsActive       bool
	Email          string
	Phone          string
	UpdatedAt      time.Time
	HasEmailChanged bool
	HasPhoneChanged bool
}

type Database struct{}

func (db *Database) Save(user *User) {
	fmt.Printf("User %s saved.\n", user.ID)
}

func (db *Database) FindUserById(id string) *User {
	return nil
}

func generateUniqueId() string {
	return fmt.Sprintf("%x", rand.Int())
}

func CreateUser(name string, age int) *User {
	user := &User{
		ID:       generateUniqueId(),
		Name:     name,
		Age:      age,
		IsActive: true,
	}
	db := &Database{}
	db.Save(user)
	return user
}

func AddUser(name string, age int) *User {
	user := &User{
		ID:       generateUniqueId(),
		Name:     name,
		Age:      age,
		IsActive: true,
	}
	db := &Database{}
	db.Save(user)
	return user
}

func UpdateUserEmail(userId string, email string) {
	db := &Database{}
	user := db.FindUserById(userId)
	if user != nil {
		user.Email = email
		user.UpdatedAt = time.Now()
		user.HasEmailChanged = true
		db.Save(user)
	}
}

func UpdateUserPhone(userId string, phone string) {
	db := &Database{}
	user := db.FindUserById(userId)
	if user != nil {
		user.Phone = phone
		user.UpdatedAt = time.Now()
		user.HasPhoneChanged = true
		db.Save(user)
	}
}

func IsAdult(age int) bool {
	return age >= 18
}

func CanVote(age int) bool {
	if age >= 18 {
		return true
	}
	return false
}

func CanDrink(age int) bool {
	if age >= 21 {
		return true
	}
	return false
}

func CalculateDiscount(price float64) float64 {
	discountRate := 0.1
	return price * discountRate
}

func CalculateTax(price float64) float64 {
	taxRate := 0.1
	return price * taxRate
}
\ No newline at end of file

File: src/typescript/structure/organisation/duplication/UserService.ts
function createUser(name: string, age: number): User {
    const user: User = {
        id: generateUniqueId(),
        name: name,
        age: age,
        isActive: true
    };
    database.save(user);
    return user;
}

function addUser(name: string, age: number): User {
    const user: User = {
        id: generateUniqueId(),
        name: name,
        age: age,
        isActive: true
    };
    database.save(user);
    return user;
}

function updateUserEmail(userId: string, email: string): void {
    const user = database.findUserById(userId);
    if (user) {
        user.email = email;
        user.updatedAt = new Date();
        user.hasEmailChanged = true;
        database.save(user);
    }
}

function updateUserPhone(userId: string, phone: string): void {
    const user = database.findUserById(userId);
    if (user) {
        user.phone = phone;
        user.updatedAt = new Date();
        user.hasPhoneChanged = true;
        database.save(user);
    }
}

function isAdult(age: number): boolean {
    return age >= 18;
}

function canVote(age: number): boolean {
    if (age >= 18) {
        return true;
    }
    return false;
}

function canDrink(age: number): boolean {
    if (age >= 21) {
        return true;
    }
    return false;
}

function calculateDiscount(price: number): number {
    const discountRate = 0.1;
    return price * discountRate;
}

function calculateTax(price: number): number {
    const taxRate = 0.1;
    return price * taxRate;
}

interface User {
    id: string;
    name: string;
    age: number;
    isActive: boolean;
    email?: string;
    phone?: string;
    updatedAt?: Date;
    hasEmailChanged?: boolean;
    hasPhoneChanged?: boolean;
}

const database = {
    save: (user: User) => {
        console.log(`User ${user.id} saved.`);
    },
    findUserById: (id: string): User | undefined => {
        return undefined;
    }
};

function generateUniqueId(): string {
    return Math.random().toString(36).substr(2, 9);
}
\ No newline at end of file